wicket-core/src/main/java/org/apache/wicket/Application.java

```diff
@@ -518,9 +519,13 @@ public void logResponseTarget(final IRequestHandler requestTarget)
 	 * @throws IllegalArgumentException
 	 * @see MetaDataKey
 	 */
-	public final synchronized <T> Application setMetaData(final MetaDataKey<T> key, final Object object)
+	@Override
+	public final <T> Application setMetaData(final MetaDataKey<T> key, final T object)
```

Comment:
I had to change the signature of this method.
Ideally, it should have been that way anyway.
Everything still compiles, which means we weren't putting values in keys that weren't expecting the value.

I fear that some users might be doing exactly that though, so this whole IMetadataContext might have to be put off for a bigger version.